rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidEditTextFactory.java

```diff
@@ -29,14 +29,15 @@
 
     @Override
     protected void attachLayout(String stepName, Context context, JsonFormFragment formFragment, JSONObject jsonObject, MaterialEditText editText, ImageView editButton) throws Exception {
-        super.attachLayout(stepName, context, formFragment, jsonObject, editText, editButton);
 
         if (!jsonObject.has(JsonFormConstants.HINT)) {
             editText.setFloatingLabel(MaterialEditText.FLOATING_LABEL_NONE);
         }
 
         updateMargin(editText, jsonObject, context);
         updateTopPadding(editText, jsonObject, context);
+
+        super.attachLayout(stepName, context, formFragment, jsonObject, editText, editButton);
```

Comment:
@syedowaisali90 what was causing the crash here?

Reply:
@vincent-karuri Actually, in the shipping preparation form, we are pre-populating the SENDER_NAME text field. Therefore, when we call the method `setText()`  then it internally triggers the listener method `afterTextChanged`.

https://github.com/rengwuxian/MaterialEditText/blob/f6e9fa42213cda552764c2a63f93a1464fa91d04/library/src/main/java/com/rengwuxian/materialedittext/MaterialEditText.java#L877-L889

So, the animator call this method internally 
```java
if (Looper.myLooper() == null) {
    throw new AndroidRuntimeException("Animators may only be run on Looper threads");
}
```
Therefore, Because the form in the loading state app will crash with the above exception. So, In order to disable the floating animation, I just swap the `super.attachLayout`. It was working fine but @mwkhan1983 shared a video that still crashing sometimes.

Reply:
idk if I fully understand how this change is disabling the animation. Where is the pre-population happening? Is it done on a background thread?

Reply:
https://github.com/rengwuxian/MaterialEditText/blob/f6e9fa42213cda552764c2a63f93a1464fa91d04/library/src/main/java/com/rengwuxian/materialedittext/MaterialEditText.java#L877-L889 if you notice in this method there is a variable called `floatingLabelEnabled`.

So, I have disabled the above variable here https://github.com/onaio/rdt-standard/blob/master/rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidEditTextFactory.java#L34-L36 before setup the super call.

Here is pre-population happening https://github.com/onaio/rdt-standard/blob/master/rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/util/CovidRDTJsonFormUtils.java#L176-L181

Reply:
This is the log of new crash
```java
Caused by: java.lang.ArrayIndexOutOfBoundsException: length=1; index=1
        at android.text.SpannableStringBuilder.getSpansRec(SpannableStringBuilder.java:987)
        at android.text.SpannableStringBuilder.getSpans(SpannableStringBuilder.java:880)
        at android.text.SpannableStringBuilder.getSpans(SpannableStringBuilder.java:849)
        at android.text.SpannableStringBuilder.<init>(SpannableStringBuilder.java:81)
        at android.text.SpannableStringBuilder.<init>(SpannableStringBuilder.java:52)
        at android.text.Editable$Factory.newEditable(Editable.java:145)
        at android.widget.TextView.setText(TextView.java:6206)
        at android.widget.TextView.setText(TextView.java:6143)
        at android.widget.EditText.setText(EditText.java:122)
        at android.widget.TextView.setText(TextView.java:6095)
        at android.widget.TextView.setTransformationMethod(TextView.java:2611)
        at android.widget.TextView.applySingleLine(TextView.java:10320)
        at android.widget.TextView.setInputType(TextView.java:6612)
        at com.vijay.jsonwizard.widgets.EditTextFactory.attachLayout(EditTextFactory.java:175)
        at io.ona.rdt.widget.CovidEditTextFactory.attachLayout(CovidEditTextFactory.java:40)
        at com.vijay.jsonwizard.widgets.EditTextFactory.attachJson(EditTextFactory.java:100)
        at io.ona.rdt.widget.CovidEditTextFactory.attachJson(CovidEditTextFactory.java:25)
        at com.vijay.jsonwizard.widgets.EditTextFactory.getViewsFromJson(EditTextFactory.java:81)
        at com.vijay.jsonwizard.interactors.JsonFormInteractor.fetchViews(JsonFormInteractor.java:223)
        at com.vijay.jsonwizard.interactors.JsonFormInteractor.fetchFields(JsonFormInteractor.java:209)
        at com.vijay.jsonwizard.interactors.JsonFormInteractor.fetchFormElements(JsonFormInteractor.java:163)
        at com.vijay.jsonwizard.presenters.JsonFormFragmentPresenter$2.run(JsonFormFragmentPresenter.java:280)
```

Reply:
@vincent-karuri can we removed the pre-populated field i.e `SENDER_NAME`

Reply:
what causes the error?

Reply:
I think that sometimes pre-filled work and sometimes not you can experience this just go to the shipping form and then go manual details and do back and forth.

Reply:
`back and forth` does this mean navigating to and from the manual details page?

Reply:
Yes

Reply:
how many times do you need to do that to get this to reproduce?

Reply:
happening randomly not the exact time.

Reply:
Unless there's a definitive way to reproduce the bug then it's not possible to find a definitive solution. From the stacktrace, the error https://github.com/onaio/rdt-standard/pull/960/files#r620212053 is from line 175 of edittextfactory. If it's possible to fix it from there then ok. I also don't understand what this error https://github.com/onaio/rdt-standard/pull/960/files#r620048768 was referring to then. Are they both occurring?

If this is only randomly and rarely occurring then I don't think we should be putting a lot of effort into this.

Reply:
only this occurring https://github.com/onaio/rdt-standard/pull/960#discussion_r620212053 randomly.

Reply:
then that can be addressed separately if need be. I still don't see how this pr addresses https://github.com/onaio/rdt-standard/pull/960#discussion_r620113087 though.

Reply:
here is a complete stacktrace.

```java
Caused by: android.util.AndroidRuntimeException: Animators may only be run on Looper threads
        at com.nineoldandroids.animation.ValueAnimator.start(ValueAnimator.java:918)
        at com.nineoldandroids.animation.ValueAnimator.start(ValueAnimator.java:951)
        at com.nineoldandroids.animation.ObjectAnimator.start(ObjectAnimator.java:385)
        at com.rengwuxian.materialedittext.MaterialEditText$2.afterTextChanged(MaterialEditText.java:886)
        at android.widget.TextView.sendAfterTextChanged(TextView.java:10652)
        at android.widget.TextView.setText(TextView.java:6324)
        at android.widget.TextView.setText(TextView.java:6143)
        at android.widget.EditText.setText(EditText.java:122)
        at android.widget.TextView.setText(TextView.java:6095)
        at android.widget.TextView.setTransformationMethod(TextView.java:2611)
        at android.widget.TextView.applySingleLine(TextView.java:10320)
        at android.widget.TextView.setInputType(TextView.java:6612)
        at com.vijay.jsonwizard.widgets.EditTextFactory.attachLayout(EditTextFactory.java:175)
        at io.ona.rdt.widget.CovidEditTextFactory.attachLayout(CovidEditTextFactory.java:32)
        at com.vijay.jsonwizard.widgets.EditTextFactory.attachJson(EditTextFactory.java:100)
        at io.ona.rdt.widget.CovidEditTextFactory.attachJson(CovidEditTextFactory.java:25)
        at com.vijay.jsonwizard.widgets.EditTextFactory.getViewsFromJson(EditTextFactory.java:81)
        at com.vijay.jsonwizard.interactors.JsonFormInteractor.fetchViews(JsonFormInteractor.java:223)
        at com.vijay.jsonwizard.interactors.JsonFormInteractor.fetchFields(JsonFormInteractor.java:209)
        at com.vijay.jsonwizard.interactors.JsonFormInteractor.fetchFormElements(JsonFormInteractor.java:163)
        at com.vijay.jsonwizard.presenters.JsonFormFragmentPresenter$2.run(JsonFormFragmentPresenter.java:280)
```

Reply:
@syedowaisali90 whatever call is failing here, make it in the foreground. similar to this https://github.com/onaio/rdt-standard/blob/b372d7c3eceb507633ad90474a6afb37a394be05/rdt-open-reader/src/main/java/io/ona/rdt/widget/RDTGpsFactory.java#L48.

Reply:
fixed.